File: protocol.js
const crypto = require('crypto');
const elliptic = require('elliptic');
const forge = require('node-forge');
const zlib = require('zlib');

// Constants
const AES_KEY_SIZE = 32; // 256-bit
const HMAC_KEY_SIZE = 32;
const NONCE_SIZE = 12;
const RSA_KEY_SIZE = 4096;
const TIMESTAMP_TOLERANCE = 30; // Seconds for replay attack protection

class Encryption {
    constructor() {
        // ECDH (Elliptic Curve Diffie-Hellman) Key Exchange
        this.ecdh = new elliptic.ec('p384');
        this.ecdhPrivateKey = this.ecdh.genKeyPair();
        this.ecdhPublicKey = this.ecdhPrivateKey.getPublic();

        // RSA (4096-bit RSA keys)
        this.rsa = forge.pki.rsa;
        this.rsaPrivateKey = this.rsa.generateKeyPair(RSA_KEY_SIZE);
        this.rsaPublicKey = this.rsaPrivateKey.publicKey;

        this.sharedSecret = null;
    }

    // Get the ECDH public key (PEM encoded)
    getEcdhPublicKey() {
        return this.ecdhPublicKey.encode('pem');
    }

    // Get the RSA public key (PEM encoded)
    getRsaPublicKey() {
        return forge.pki.publicKeyToPem(this.rsaPublicKey);
    }

    // Derive shared secret using ECDH
    deriveSharedSecret(peerPublicKeyPem) {
        const peerPublicKey = this.ecdh.keyFromPublic(peerPublicKeyPem, 'pem');
        const sharedSecret = this.ecdhPrivateKey.derive(peerPublicKey.getPublic());
        
        // Use HKDF to derive AES and HMAC keys
        const hkdf = crypto.createHmac('sha256', sharedSecret.toString(16));
        hkdf.update('E2EE Key Derivation');
        const keyMaterial = hkdf.digest();

        this.aesKey = keyMaterial.slice(0, AES_KEY_SIZE);
        this.hmacKey = keyMaterial.slice(AES_KEY_SIZE, AES_KEY_SIZE + HMAC_KEY_SIZE);

        return this.aesKey.toString('base64');
    }

    // Sign the message using RSA
    signMessage(message) {
        const md = forge.md.sha256.create();
        md.update(message);
        const signature = this.rsaPrivateKey.sign(md);
        return forge.util.encode64(signature);
    }

    // Verify the signature using RSA
    verifySignature(message, signature) {
        const md = forge.md.sha256.create();
        md.update(message);
        const decodedSignature = forge.util.decode64(signature);
        return this.rsaPublicKey.verify(md.digest().bytes(), decodedSignature);
    }

    // Encrypt the plaintext message using AES-256-GCM
    encrypt(plaintext) {
        const compressedData = zlib.deflateSync(plaintext);
        const nonce = crypto.randomBytes(NONCE_SIZE);
        const cipher = crypto.createCipheriv('aes-256-gcm', this.aesKey, nonce);
        const ciphertext = Buffer.concat([cipher.update(compressedData), cipher.final()]);
        const tag = cipher.getAuthTag();

        // Create HMAC for integrity
        const mac = crypto.createHmac('sha256', this.hmacKey).update(ciphertext).digest();

        // Include timestamp for replay protection
        const timestamp = Buffer.alloc(8);
        timestamp.writeBigUInt64BE(BigInt(Math.floor(Date.now() / 1000)), 0);

        return Buffer.concat([nonce, tag, timestamp, mac, ciphertext]).toString('base64');
    }

    // Decrypt the encrypted data using AES-256-GCM
    decrypt(encryptedData) {
        const data = Buffer.from(encryptedData, 'base64');
        const nonce = data.slice(0, NONCE_SIZE);
        const tag = data.slice(NONCE_SIZE, NONCE_SIZE + 16);
        const timestamp = data.slice(NONCE_SIZE + 16, NONCE_SIZE + 24);
        const mac = data.slice(NONCE_SIZE + 24, NONCE_SIZE + 56);
        const ciphertext = data.slice(NONCE_SIZE + 56);

        // Replay protection: check timestamp
        if (Math.abs(Date.now() / 1000 - timestamp.readBigUInt64BE(0)) > TIMESTAMP_TOLERANCE) {
            throw new Error('Replay attack detected!');
        }

        // Verify HMAC
        const expectedMac = crypto.createHmac('sha256', this.hmacKey).update(ciphertext).digest();
        if (!crypto.timingSafeEqual(mac, expectedMac)) {
            throw new Error('Data integrity compromised!');
        }

        const decipher = crypto.createDecipheriv('aes-256-gcm', this.aesKey, nonce);
        decipher.setAuthTag(tag);
        const decryptedData = Buffer.concat([decipher.update(ciphertext), decipher.final()]);

        // Decompress the data after decryption
        return zlib.inflateSync(decryptedData).toString();
    }

    // Encrypt the symmetric AES key using RSA
    encryptSymmetricKey() {
        return forge.util.encode64(this.rsaPublicKey.encrypt(this.aesKey));
    }

    // Decrypt the symmetric AES key using RSA
    decryptSymmetricKey(encryptedKey) {
        const decodedKey = forge.util.decode64(encryptedKey);
        this.aesKey = this.rsaPrivateKey.decrypt(decodedKey);
        return this.aesKey.toString('base64');
    }
}

module.exports = Encryption;

File: protocol.py
import os
import time
import struct
import hashlib
import base64
import json
from cryptography.hazmat.primitives.asymmetric import rsa, ec, padding
from cryptography.hazmat.primitives import serialization, hashes
from cryptography.hazmat.primitives.kdf.hkdf import HKDF
from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes
import hmac

# Constants
AES_KEY_SIZE = 32  # 256-bit
HMAC_KEY_SIZE = 32
NONCE_SIZE = 12
RSA_KEY_SIZE = 4096
TIMESTAMP_TOLERANCE = 30  # Seconds for replay attack protection

class main:
    def __init__(self):
        """Initialize E2EE with ECDH key exchange and RSA encryption."""
        self.ecdh_private_key = ec.generate_private_key(ec.SECP384R1())  # Use a more secure curve
        self.ecdh_public_key = self.ecdh_private_key.public_key()
        self.rsa_private_key = rsa.generate_private_key(public_exponent=65537, key_size=RSA_KEY_SIZE)
        self.rsa_public_key = self.rsa_private_key.public_key()

    def get_ecdh_public_key(self):
        """Export ECDH public key."""
        return self.ecdh_public_key.public_bytes(
            encoding=serialization.Encoding.PEM,
            format=serialization.PublicFormat.SubjectPublicKeyInfo
        )

    def get_rsa_public_key(self):
        """Export RSA public key."""
        return self.rsa_public_key.public_bytes(
            encoding=serialization.Encoding.PEM,
            format=serialization.PublicFormat.SubjectPublicKeyInfo
        )

    def derive_shared_secret(self, peer_public_key_pem):
        """Derive shared secret using ECDH."""
        peer_public_key = serialization.load_pem_public_key(peer_public_key_pem)
        shared_secret = self.ecdh_private_key.exchange(ec.ECDH(), peer_public_key)

        # Use HKDF to derive AES and HMAC keys
        kdf = HKDF(algorithm=hashes.SHA256(), length=AES_KEY_SIZE + HMAC_KEY_SIZE, salt=None, info=b'E2EE Key Derivation')
        key_material = kdf.derive(shared_secret)

        self.aes_key = key_material[:AES_KEY_SIZE]
        self.hmac_key = key_material[AES_KEY_SIZE:]
        return base64.b64encode(self.aes_key).decode()

    def sign_message(self, message):
        """Sign the message using RSA to provide authenticity."""
        signature = self.rsa_private_key.sign(
            message.encode(),
            padding.PSS(mgf=padding.MGF1(algorithm=hashes.SHA256()), salt_length=padding.PSS.MAX_LENGTH),
            hashes.SHA256()
        )
        return base64.b64encode(signature).decode()

    def verify_signature(self, message, signature):
        """Verify the signature of a message using RSA."""
        try:
            self.rsa_public_key.verify(
                base64.b64decode(signature),
                message.encode(),
                padding.PSS(mgf=padding.MGF1(algorithm=hashes.SHA256()), salt_length=padding.PSS.MAX_LENGTH),
                hashes.SHA256()
            )
            return True
        except Exception:
            return False

    def encrypt(self, plaintext):
        """Encrypts data using AES-256-GCM with integrity protection."""
        nonce = os.urandom(NONCE_SIZE)
        cipher = Cipher(algorithms.AES(self.aes_key), modes.GCM(nonce))
        encryptor = cipher.encryptor()
        ciphertext = encryptor.update(plaintext.encode()) + encryptor.finalize()

        # Create HMAC for integrity
        mac = hmac.new(self.hmac_key, ciphertext, hashlib.sha256).digest()

        # Include timestamp for replay protection
        timestamp = struct.pack(">Q", int(time.time()))

        return base64.b64encode(nonce + encryptor.tag + timestamp + mac + ciphertext).decode()

    def decrypt(self, encrypted_data):
        """Decrypts data and verifies integrity & freshness."""
        data = base64.b64decode(encrypted_data)

        nonce = data[:NONCE_SIZE]
        tag = data[NONCE_SIZE:NONCE_SIZE+16]
        timestamp = struct.unpack(">Q", data[NONCE_SIZE+16:NONCE_SIZE+24])[0]
        mac = data[NONCE_SIZE+24:NONCE_SIZE+56]
        ciphertext = data[NONCE_SIZE+56:]

        # Replay protection: check timestamp
        if abs(time.time() - timestamp) > TIMESTAMP_TOLERANCE:
            raise Exception("Replay attack detected!")

        # Verify HMAC
        expected_mac = hmac.new(self.hmac_key, ciphertext, hashlib.sha256).digest()
        if not hmac.compare_digest(mac, expected_mac):
            raise Exception("Data integrity compromised!")

        # Decrypt
        cipher = Cipher(algorithms.AES(self.aes_key), modes.GCM(nonce, tag))
        decryptor = cipher.decryptor()
        return decryptor.update(ciphertext) + decryptor.finalize()

    def encrypt_symmetric_key(self):
        """Encrypt AES key with RSA-4096."""
        encrypted_key = self.rsa_public_key.encrypt(
            self.aes_key,
            padding.OAEP(mgf=padding.MGF1(algorithm=hashes.SHA256()), algorithm=hashes.SHA256(), label=None)
        )
        return base64.b64encode(encrypted_key).decode()

    def decrypt_symmetric_key(self, encrypted_key):
        """Decrypt AES key using RSA-4096."""
        decrypted_key = self.rsa_private_key.decrypt(
            base64.b64decode(encrypted_key),
            padding.OAEP(mgf=padding.MGF1(algorithm=hashes.SHA256()), algorithm=hashes.SHA256(), label=None)
        )
        self.aes_key = decrypted_key
        return base64.b64encode(self.aes_key).decode()

    def encrypt_metadata(self, metadata):
        """Encrypts metadata (IP, port, etc.)."""
        metadata_json = json.dumps(metadata).encode()
        cipher = Cipher(algorithms.AES(self.aes_key), modes.GCM(os.urandom(NONCE_SIZE)))
        encryptor = cipher.encryptor()
        encrypted_metadata = encryptor.update(metadata_json) + encryptor.finalize()
        return base64.b64encode(encrypted_metadata).decode()

    def decrypt_metadata(self, encrypted_metadata):
        """Decrypts metadata."""
        encrypted_metadata = base64.b64decode(encrypted_metadata)
        cipher = Cipher(algorithms.AES(self.aes_key), modes.GCM(os.urandom(NONCE_SIZE)))
        decryptor = cipher.decryptor()
        return json.loads(decryptor.update(encrypted_metadata) + decryptor.finalize())
    
    def handshake(self, peer_public_key_pem):
        """Complete the handshake protocol."""
        # Generate and exchange public keys, and derive shared secret
        peer_public_key = serialization.load_pem_public_key(peer_public_key_pem)
        shared_secret = self.derive_shared_secret(peer_public_key_pem)

        # Return the shared secret and the public key for the peer to encrypt the symmetric key
        return {
            "shared_secret": shared_secret,
            "public_key": self.get_rsa_public_key(),
        }

    def secure_message_exchange(self, peer_public_key_pem, message):
        """Exchange a secure message (sign, encrypt, decrypt)."""
        # Handshake to derive shared secret and public keys
        handshake_data = self.handshake(peer_public_key_pem)
        
        # Sign the message before encrypting
        signed_message = self.sign_message(message)
        
        # Encrypt the message
        encrypted_message = self.encrypt(message)
        
        return {
            "signed_message": signed_message,
            "encrypted_message": encrypted_message,
            "handshake_data": handshake_data
        }

    def authenticate_peer(self, peer_public_key_pem, signed_message):
        """Authenticate the peer by verifying their signed message."""
        if self.verify_signature(signed_message, peer_public_key_pem):
            return True
        return False
